In [2]:
import pandas as pd
import numpy as np

In [76]:
df = pd.read_csv('ionosphere.data')
tf = df[300:351]
df = df[0:300]
tf

1  0  0.99539  -0.05889  0.85243  0.02306  0.83398  -0.37708      1.1  \
300  1  0  0.83367   0.31456  0.65541  0.57671  0.34962   0.70677  0.17293   
301  1  0  0.98455  -0.02736  0.98058 -0.04104  1.00000  -0.07635  0.98720   
302  1  0  0.94052  -0.01531  0.94170  0.01001  0.94994  -0.01472  0.95878   
303  1  0  0.97032  -0.14384  0.91324 -0.00228  0.96575  -0.17123  0.98630   
304  1  0  0.52542  -0.03390  0.94915  0.08475  0.52542  -0.16949  0.30508   
305  1  0  0.33333  -0.25000  0.44444  0.22222  0.38889   0.16667  0.41667   
306  1  0  0.51207   1.00000  1.00000  0.53810  0.71178   0.80833  0.45622   
307  1  0  0.75564   0.49638  0.83550  0.54301  0.54916   0.72063  0.35225   
308  1  0  0.83789   0.42904  0.72113  0.58385  0.45625   0.78115  0.16470   
309  1  0  0.93658   0.35107  0.75254  0.65640  0.45571   0.88576  0.15323   
310  1  0  0.92436   0.36924  0.71976  0.68420  0.29303   0.94078 -0.11108   
311  1  0  1.00000   0.16195  1.00000 -0.05558  1.00000   0.01373  1.00000   
312  1  0  0.95559  -0.00155  0.86421 -0.13244  0.94982  -0.00461  0.82809   
313  1  0  0.79378   0.29492  0.64064  0.52312  0.41319   0.68158  0.14177   
314  1  0  0.88085   0.35232  0.68389  0.65128  0.34816   0.79784  0.05832   
315  1  0  0.89589   0.39286  0.66129  0.71804  0.29521   0.90824 -0.04787   
316  1  0  1.00000  -0.24051  1.00000 -0.20253  0.87342  -0.10127  0.88608   
317  1  0  0.74790   0.00840  0.83312  0.01659  0.82638   0.02469  0.86555   
318  1  0  0.85013   0.01809  0.92211  0.01456  0.92046   0.02180  0.92765   
319  1  0  1.00000  -0.01179  1.00000 -0.00343  1.00000  -0.01565  1.00000   
320  1  0  0.89505  -0.03168  0.87525  0.05545  0.89505   0.01386  0.92871   
321  1  0  0.90071   0.01773  1.00000 -0.01773  0.90071   0.00709  0.84752   
322  1  0  0.39394  -0.24242  0.62655  0.01270  0.45455   0.09091  0.63636   
323  1  0  0.86689   0.35950  0.72014  0.66667  0.37201   0.83049  0.08646   
324  1  0  0.89563   0.37917  0.67311  0.69438  0.35916   0.88696 -0.04193   
325  1  0  0.90547   0.41113  0.65354  0.74761  0.29921   0.95905 -0.13342   
326  1  0  1.00000   1.00000  0.36700  0.06158  0.12993   0.92713 -0.27586   
327  1  0  1.00000   0.51515  0.45455  0.33333  0.06061   0.36364 -0.32104   
328  1  0  0.88110   0.00000  0.94817 -0.02744  0.93598  -0.01220  0.90244   
329  1  0  0.82624   0.08156  0.79078 -0.08156  0.90426  -0.01773  0.92908   
330  1  0  0.74468   0.10638  0.88706  0.00982  0.88542   0.01471  0.87234   
331  1  0  0.87578   0.03727  0.89951  0.00343  0.89210   0.00510  0.86335   
332  1  0  0.97513   0.00710  0.98579  0.01954  1.00000   0.01954  0.99290   
333  1  0  1.00000   0.01105  1.00000  0.01105  1.00000   0.02320  0.99448   
334  1  0  1.00000  -0.01342  1.00000  0.01566  1.00000  -0.00224  1.00000   
335  1  0  0.88420   0.36724  0.67123  0.67382  0.39613   0.86399  0.02424   
336  1  0  0.90147   0.41786  0.64131  0.75725  0.30440   0.95148 -0.20449   
337  1  0  0.32789   0.11042  0.15970  0.29308  0.14020   0.74485 -0.25131   
338  1  0  0.65845   0.43617  0.44681  0.74804  0.05319   0.85106 -0.32027   
339  1  0  0.19466   0.05725  0.04198  0.25191 -0.10557   0.48866 -0.18321   
340  1  0  0.98002   0.00075  1.00000  0.00000  0.98982  -0.00075  0.94721   
341  1  0  0.82254  -0.07572  0.80462  0.00231  0.87514  -0.01214  0.86821   
342  1  0  0.35346  -0.13768  0.69387 -0.02423  0.68195  -0.03574  0.55717   
343  1  0  0.76046   0.01092  0.86335  0.00258  0.85821   0.00384  0.79988   
344  1  0  0.66667  -0.01366  0.97404  0.06831  0.49590   0.50137  0.75683   
345  1  0  0.83508   0.08298  0.73739 -0.14706  0.84349  -0.05567  0.90441   
346  1  0  0.95113   0.00419  0.95183 -0.02723  0.93438  -0.01920  0.94590   
347  1  0  0.94701  -0.00034  0.93207 -0.03227  0.95177  -0.03431  0.95584   
348  1  0  0.90608  -0.01657  0.98122 -0.01989  0.95691  -0.03646  0.85746   
349  1  0  0.84710   0.13533  0.73638 -0.06151  0.87873   0.08260  0.88928   

     0.03760  ...  -0.5117

In [75]:
np.size(df, 1)

300 35 50 35 -1


In [121]:
#### ==========Part 1===========#

import cvxpy as cp

'''Training'''
df_np = df.to_numpy()
for i in range(0,np.size(df_np,0)):
    if(df_np[i][34]=='b'):
        df_np[i][34] = -1
    elif(df_np[i][34]=='g'):
        df_np[i][34] = 1
n = 34

Y = [df_np[:,34]]
X = df_np[:,0:34]
beta = cp.Variable(1)
w = cp.Variable((34,1))

cost = cp.sum_squares(X@w + beta - Y)
res = cp.Problem(cp.Minimize(cost)).solve()


'''Testing'''
tf_np = tf.to_numpy()
loss = 0

for i in range(np.size(tf_np,0)):
        if(tf_np[i][34]=='b'):
            tf_np[i][34] = -1
        elif(tf_np[i][34]=='g'):
            tf_np[i][34] = 1

for i in range((np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    if((res[0]>=0 and expected>=0) or (res[0]<0 and expected<0)): pass
    else: loss+=1
    #loss += abs(round(res[0],0)-expected)

print(1-loss/51)

1.0


In [120]:
#==========Part 2===========#


import cvxpy as cp

'''Training'''
n = 34

Y = [df_np[:,34]]
Y = np.asarray(Y)
X = df_np[:,0:34]
beta = cp.Variable(1)
w = cp.Variable((34,1))

log_likelihood = cp.sum(cp.logistic(-Y@(X@w + beta)))
#cost = cp.sum(cp.huber(X@w + beta - Y, 1))
res = cp.Problem(cp.Minimize(log_likelihood)).solve()


'''Testing'''
loss = 0

for i in range(0, (np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    if((res[0]>=0 and expected>=0) or (res[0]<0 and expected<0)): pass
    else: loss = loss+1
    #loss += abs(round(res[0])-expected)

print(1-loss/51)

0.7647058823529411


In [72]:
print(np.size(df_np,0),np.size(df_np,1),np.size(tf_np,0),np.size(tf_np,1),df_np[0][34])

300 35 50 35 -1


In [122]:
#==========Part 3===========#


import cvxpy as cp

'''Training'''
Y = [df_np[:,34]]
Y = np.asarray(Y)
X = df_np[:,0:34]
beta = cp.Variable(1)
w = cp.Variable((34,1))

cost = cp.sum(cp.maximum(0,1-Y@(X@w + beta)))
res = cp.Problem(cp.Minimize(cost)).solve()

'''Testing'''
loss = 0

for i in range(0, (np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    #print(res[0], expected)
    if((res[0]>=0 and expected>=0) or (res[0]<0 and expected<0)): pass
    else: loss = loss+1
    
    #loss += abs(res[0]-expected)
print(1-loss/51)

0.7450980392156863


Results:
So as you can see the accuracies are-
A. 100%
B. 76.47058823529411 %
C. 74.50980392156863 %